### PR TITLE
Fix #2421: Add errors for inline on non supported trees

### LINKED ISSUE #2421: Missing errors for `inline var`, `inline object`, `inline class` and `inline type`.
* [x] `inline var foo`: Should not be allowed, it will break after the first assignment. Possibly for `lazy val` too.
* [x] `inline class Foo`: Should not be allowed.
* [x] `inline object Foo`: Should have a better error message. Currently _right-hand side of inline object Foo must be a constant expression or a function_.
* [x] `inline type Foo`: Should not be allowed, it makes not sense.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by felixmulder** [comment]
@nicolasstucki - needs rebase

File: compiler/src/dotty/tools/dotc/typer/Checking.scala [hunk 1]
```diff
@@ -550,6 +550,28 @@ trait Checking {
...
   }
 
+  /** Check that inline is only used on valid trees */
+  def checkInlineKeyword(tree: tpd.Tree)(implicit ctx: Context) = {
+    def errorOn(what: String) =
+      ctx.error("inlined keyword cannot be used on " + what, tree.pos)
+    val sym = tree.symbol
+    if (sym is Inline) {
+      tree match {
+        case _: TypeDef =>
```
Comment (DarkDimius) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala]: PlainPrinter.kindString

**Comment by nicolasstucki** [comment]
Moved checks to checkFlags

**Comment by nicolasstucki** [comment]
All requested changes were addressed.

**Comment by nicolasstucki** [comment]
@odersky it looks like your commits lost the test file https://github.com/dotty-staging/dotty/blob/ee331a6fa255638c8db59dae269198d3a8c3dba9/tests/neg/i2421.scala

**Comment by odersky** [comment]
I found a solution that requires less code and fits better. @nicolasstucki do you want to review?

File: compiler/src/dotty/tools/dotc/typer/Checking.scala [hunk 2]
```diff
@@ -315,7 +315,10 @@ object Checking {
         fail(AbstractMemberMayNotHaveModifier(sym, flag))
     def checkNoConflict(flag1: FlagSet, flag2: FlagSet) =
       if (sym.is(allOf(flag1, flag2)))
-        fail(i"illegal combination of modifiers: $flag1 and $flag2 for: $sym")
+        fail(i"illegal combination of modifiers: `$flag1` and `$flag2` for: $sym")
+    def checkApplicable(flag: FlagSet, ok: Boolean) =
+      if (!ok && !sym.is(Synthetic))
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala]: Why is the `!sym.is(Synthetic)` needed here? Is it only for the `lazy val` of the modules?

Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala]: For now yes, but there might be other synthetic symbols with conflicts. In any case, if it is synthetic we should not issue an error message to the user.

Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala]: Ok

**Comment by nicolasstucki** [comment]
LGTM

**Pull request merged**